### PR TITLE
Fix nadam optimizer

### PR DESCRIPTION
**Pull request by strubell**
Resolves issues #15035 and #13980.

### CODE REVIEW

**Comment by case540** [comment]
Thanks for the PR! Could you add a test that would have failed without this change.

**Comment by strubell** [comment]
Thanks for the review! I updated the test. 

The test wasn't failing before because the "sparse" test was creating a sparse vector that was actually dense (i.e. it's of size 2 and values are being specified for both indices). FWIW it looks like a number of other optimizers (at least Adam, LazyAdam) have the same shortcoming in their tests.

**Comment by josh11b** [comment]
Alex, it looks like you reviewed the original nadam PR https://github.com/tensorflow/tensorflow/pull/9889. I'm not familiar with the nadam algorithm.

**Review by alextp — DISMISSED** [review]

**Comment by alextp** [comment]
Hey Emma!

**Comment by strubell** [comment]
Hey Alex, thanks for the review!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@strubell could you pull rebase and push again?

**Comment by strubell** [comment]
Sorry I've been travelling -- will do this soon. Thanks for taking a look!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/contrib/opt/python/training/nadam_optimizer_test.py [hunk 1]
```diff
@@ -52,36 +52,43 @@ def nadam_update_numpy(param,
 class NadamOptimizerTest(test.TestCase):
 
   def doTestSparse(self, use_resource=False):
+    
+    # need to use a larger value of epsilon here so that
+    # np.sqrt(v_t) + epsilon doesn't get rounded to 0 when 
+    # the dtype is half and np.sqrt(v_t) = 0, as is the case
+    # when the gradient is 0 
+    sparse_epsilon = 1e-7
```
Comment (tanzhenyu) [inline tensorflow/contrib/opt/python/training/nadam_optimizer_test.py]: I see the reasoning. Why won't it happen for dense?

Comment (tanzhenyu) [reply tensorflow/contrib/opt/python/training/nadam_optimizer_test.py]: I mean since v_t is initialized to be 0, so sqrt(v_t) is 0 as well, if epsilon also gets rounded to 0 because of half float, aren't we dividing by 0 no matter dense or sparse?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @case540: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by tanzhenyu — APPROVED** [review]

**Comment by rthadur** [comment]
@strubell could you pull rebase and push again?

**Comment by rthadur** [comment]
Please send this PR to tensorflow/addons instead; contrib is deprecated and stated to be removed so we're not making new changes to it.
